File: smartboard/src/core/analysis.ts
/**
 * 高级分析算法 — 从 dashboard_gen.py 迁移
 * timeseries / decile / cluster (K-means)
 */

// ====== Timeseries ======

export interface TimeseriesData {
  labels: string[]
  values: number[]
  ma: (number | null)[]       // MA3 移动平均
  mom: (number | null)[]       // 环比 %
  yoy: (number | null)[]       // 同比 % (仅月度)
  trend: number[]              // 线性回归趋势
  forecast: { labels: string[]; values: number[] }
}

function getNumericVal(v: string | number | undefined): number {
  if (v === undefined || v === null || v === '') return NaN
  if (typeof v === 'number') return v
  let s = String(v).trim()
  if (s.endsWith('%')) s = s.slice(0, -1)
  s = s.replace(/,/g, '')
  const n = parseFloat(s)
  return isNaN(n) ? NaN : n
}

function getPeriodKey(dateStr: string, period: 'month' | 'quarter' | 'year'): string | null {
  const m = dateStr.match(/^(\d{4})[-/.](\d{1,2})/)
  if (!m) return null
  const y = m[1], mo = parseInt(m[2])
  if (period === 'year') return y
  if (period === 'quarter') return `${y}-Q${Math.ceil(mo / 3)}`
  return `${y}-${String(mo).padStart(2, '0')}`
}

function nextPeriodKeys(lastKey: string, period: 'month' | 'quarter' | 'year', count: number): string[] {
  const result: string[] = []
  if (period === 'year') {
    let y = parseInt(lastKey)
    for (let i = 0; i < count; i++) result.push(String(++y))
  } else if (period === 'quarter') {
    const [y, q] = lastKey.split('-Q').map(Number)
    let yr = y, qt = q
    for (let i = 0; i < count; i++) {
      qt++
      if (qt > 4) { qt = 1; yr++ }
      result.push(`${yr}-Q${qt}`)
    }
  } else {
    const [y, m] = lastKey.split('-').map(Number)
    let yr = y, mo = m
    for (let i = 0; i < count; i++) {
      mo++
      if (mo > 12) { mo = 1; yr++ }
      result.push(`${yr}-${String(mo).padStart(2, '0')}`)
    }
  }
  return result
}

/**
 * 时序分析：按周期聚合，计算 MA3、环比、同比、趋势、预测
 */
export function computeTimeseries(
  rows: Record<string, string | number>[],
  dateCol: string,
  metricCol: string,
  period: 'month' | 'quarter' | 'year' = 'month',
): TimeseriesData | null {
  // 按周期分组求和
  const groups: Record<string, number> = {}
  for (const row of rows) {
    const dv = String(row[dateCol] || '').trim()
    const key = getPeriodKey(dv, period)
    if (!key) continue
    const v = getNumericVal(row[metricCol])
    if (!isNaN(v)) groups[key] = (groups[key] || 0) + v
  }

  const sortedKeys = Object.keys(groups).sort()
  if (sortedKeys.length < 2) return null

  const labels = sortedKeys
  const values = sortedKeys.map((k) => groups[k])

  // MA3 移动平均
  const ma: (number | null)[] = values.map((_, i) => {
    if (i < 2) return null
    return (values[i - 2] + values[i - 1] + values[i]) / 3
  })

  // 环比 MoM%
  const mom: (number | null)[] = values.map((v, i) => {
    if (i === 0 || values[i - 1] === 0) return null
    return ((v - values[i - 1]) / Math.abs(values[i - 1])) * 100
  })

  // 同比 YoY% — 仅月度数据
  let yoy: (number | null)[] = []
  if (period === 'month' && labels.length > 12) {
    yoy = values.map((v, i) => {
      if (i < 12 || values[i - 12] === 0) return null
      return ((v - values[i - 12]) / Math.abs(values[i - 12])) * 100
    })
  } else {
    yoy = values.map(() => null)
  }

  // 线性回归趋势
  const n = values.length
  const xs = values.map((_, i) => i)
  const meanX = xs.reduce((a, b) => a + b, 0) / n
  const meanY = values.reduce((a, b) => a + b, 0) / n
  let num = 0, den = 0
  for (let i = 0; i < n; i++) {
    num += (xs[i] - meanX) * (values[i] - meanY)
    den += (xs[i] - meanX) ** 2
  }
  const slope = den !== 0 ? num / den : 0
  const intercept = meanY - slope * meanX
  const trend = values.map((_, i) => intercept + slope * i)

  // 预测：外推 3 个周期
  const forecastCount = 3
  const forecastLabels = nextPeriodKeys(labels[labels.length - 1], period, forecastCount)
  const forecastValues = forecastLabels.map((_, i) => intercept + slope * (n + i))

  return { labels, values, ma, mom, yoy, trend, forecast: { labels: forecastLabels, values: forecastValues } }
}

// ====== Decile (十分位) ======

export interface DecileData {
  labels: string[]   // D1..D10
  counts: number[]
  sums: number[]
  avgs: number[]
  ranges: string[]
}

function fmtCompact(n: number): string {
  const a = Math.abs(n)
  if (a >= 1e8) return (n / 1e8).toFixed(1) + '亿'
  if (a >= 1e4) return (n / 1e4).toFixed(1) + '万'
  return n.toLocaleString('zh-CN', { maximumFractionDigits: 1 })
}

/**
 * 十分位分析：按指标排序后等分为 10 个桶
 */
export function computeDeciles(
  rows: Record<string, string | number>[],
  metricCol: string,
): DecileData | null {
  const nums = rows
    .map((r) => getNumericVal(r[metricCol]))
    .filter((n) => !isNaN(n))
    .sort((a, b) => a - b)

  if (nums.length < 10) return null

  const bucketSize = Math.floor(nums.length / 10)
  const labels: string[] = []
  const counts: number[] = []
  const sums: number[] = []
  const avgs: number[] = []
  const ranges: string[] = []

  for (let i = 0; i < 10; i++) {
    const start = i * bucketSize
    const end = i === 9 ? nums.length : (i + 1) * bucketSize
    const bucket = nums.slice(start, end)
    const sum = bucket.reduce((a, b) => a + b, 0)

    labels.push(`D${i + 1}`)
    counts.push(bucket.length)
    sums.push(sum)
    avgs.push(bucket.length > 0 ? sum / bucket.length : 0)
    ranges.push(`${fmtCompact(bucket[0])}–${fmtCompact(bucket[bucket.length - 1])}`)
  }

  return { labels, counts, sums, avgs, ranges }
}

// ====== K-means Clustering ======

export interface ClusterData {
  points: { x: number; y: number; cluster: number; label: string }[]
  centroids: { x: number; y: number }[]
  colX: string
  colY: string
}

/**
 * K-means 聚类（20 轮迭代，确定性初始化）
 */
export function computeClusters(
  rows: Record<string, string | number>[],
  metricCols: string[],
  k: number = 3,
  xCol?: string,
  yCol?: string,
): ClusterData | null {
  const colX = xCol || metricCols[0]
  const colY = yCol || metricCols[1] || metricCols[0]
  if (!colX || !colY) return null

  // 提取有效点
  const points: { x: number; y: number; label: string }[] = []
  for (const row of rows) {
    const x = getNumericVal(row[colX])
    const y = getNumericVal(row[colY])
    if (!isNaN(x) && !isNaN(y)) {
      // 取第一列文本作为 label
      const label = String(Object.values(row)[0] ?? '')
      points.push({ x, y, label })
    }
  }
  if (points.length < k) return null

  // 初始化聚类中心（等间距取样）
  let centroids: { x: number; y: number }[] = []
  const step = Math.floor(points.length / k)
  for (let i = 0; i < k; i++) {
    const idx = Math.min(i * step, points.length - 1)
    centroids.push({ x: points[idx].x, y: points[idx].y })
  }

  // 迭代
  let assignments = new Array(points.length).fill(0)
  for (let iter = 0; iter < 20; iter++) {
    // 分配
    const newAssign = points.map((p) => {
      let minDist = Infinity, minC = 0
      for (let c = 0; c < centroids.length; c++) {
        const d = (p.x - centroids[c].x) ** 2 + (p.y - centroids[c].y) ** 2
        if (d < minDist) { minDist = d; minC = c }
      }
      return minC
    })

    // 检查收敛
    if (newAssign.every((v, i) => v === assignments[i])) break
    assignments = newAssign

    // 更新中心
    const sums = centroids.map(() => ({ x: 0, y: 0, count: 0 }))
    for (let i = 0; i < points.length; i++) {
      const c = assignments[i]
      sums[c].x += points[i].x
      sums[c].y += points[i].y
      sums[c].count++
    }
    centroids = sums.map((s, i) =>
      s.count > 0
        ? { x: s.x / s.count, y: s.y / s.count }
        : centroids[i],
    )
  }

  return {
    points: points.map((p, i) => ({ ...p, cluster: assignments[i] })),
    centroids,
    colX,
    colY,
  }
}

File: smartboard/src-tauri/src/commands/analysis.rs
// src-tauri/src/commands/analysis.rs
// 高级分析命令 — 时序 / 十分位 / K-means 聚类
// 在 spawn_blocking 中执行，不阻塞 async 线程

use crate::types::ApiResult;
use serde::Serialize;
use std::collections::HashMap;

// ──────────────── 数据结构 ────────────────

#[derive(Debug, Serialize)]
pub struct TimeseriesResult {
    pub labels: Vec<String>,
    pub values: Vec<f64>,
    pub ma: Vec<Option<f64>>,
    pub mom: Vec<Option<f64>>,
    pub yoy: Vec<Option<f64>>,
    pub trend: Vec<f64>,
    pub forecast_labels: Vec<String>,
    pub forecast_values: Vec<f64>,
}

#[derive(Debug, Serialize)]
pub struct DecileResult {
    pub labels: Vec<String>,
    pub counts: Vec<usize>,
    pub sums: Vec<f64>,
    pub avgs: Vec<f64>,
    pub ranges: Vec<String>,
}

#[derive(Debug, Serialize)]
pub struct ClusterPoint {
    pub x: f64,
    pub y: f64,
    pub cluster: usize,
    pub label: String,
}

#[derive(Debug, Serialize)]
pub struct ClusterResult {
    pub points: Vec<ClusterPoint>,
    pub centroids: Vec<(f64, f64)>,
    pub col_x: String,
    pub col_y: String,
}

// ──────────────── 辅助函数 ────────────────

fn parse_row_value(row: &HashMap<String, serde_json::Value>, col: &str) -> Option<f64> {
    row.get(col).and_then(|v| {
        if v.is_null() {
            None
        } else if let Some(n) = v.as_f64() {
            Some(n)
        } else if let Some(s) = v.as_str() {
            s.replace(',', "")
                .replace('%', "")
                .trim()
                .parse::<f64>()
                .ok()
        } else {
            None
        }
    })
}

fn get_period_key(date_str: &str, period: &str) -> Option<String> {
    let s = date_str.trim();
    if s.len() < 7 {
        return None;
    }
    let y = s[..4].parse::<i32>().ok()?;
    let mo = s[5..7].parse::<i32>().ok()?;
    match period {
        "year" => Some(format!("{}", y)),
        "quarter" => Some(format!("{}-Q{}", y, (mo + 2) / 3)),
        _ => Some(format!("{}-{:02}", y, mo)),
    }
}

fn next_period_keys(last: &str, period: &str, count: usize) -> Vec<String> {
    let mut result = Vec::new();
    let parts: Vec<&str> = last.split('-').collect();
    if period == "year" {
        let mut y: i32 = last.parse().unwrap_or(0);
        for _ in 0..count {
            y += 1;
            result.push(format!("{}", y));
        }
    } else if period == "quarter" {
        let y = parts[0].parse::<i32>().unwrap_or(0);
        let q_part = parts.last().unwrap_or(&"1");
        let q = q_part.trim_start_matches("Q").parse::<i32>().unwrap_or(1);
        let mut yr = y;
        let mut qt = q;
        for _ in 0..count {
            qt += 1;
            if qt > 4 {
                qt = 1;
                yr += 1;
            }
            result.push(format!("{}-Q{}", yr, qt));
        }
    } else {
        let y = parts[0].parse::<i32>().unwrap_or(0);
        let m = parts[1].parse::<i32>().unwrap_or(1);
        let mut yr = y;
        let mut mo = m;
        for _ in 0..count {
            mo += 1;
            if mo > 12 {
                mo = 1;
                yr += 1;
            }
            result.push(format!("{}-{:02}", yr, mo));
        }
    }
    result
}

fn fmt_compact(n: f64) -> String {
    let a = n.abs();
    if a >= 1e8 {
        format!("{:.1}亿", n / 1e8)
    } else if a >= 1e4 {
        format!("{:.1}万", n / 1e4)
    } else {
        format!("{:.1}", n)
    }
}

// ──────────────── Tauri commands ────────────────

#[tauri::command]
pub async fn compute_timeseries(
    rows_json: String,
    date_col: String,
    metric_col: String,
    period: String,
) -> ApiResult<TimeseriesResult> {
    tokio::task::spawn_blocking(move || {
        let rows: Vec<HashMap<String, serde_json::Value>> = match serde_json::from_str(&rows_json) {
            Ok(r) => r,
            Err(e) => return ApiResult::failure(format!("JSON解析失败: {e}")),
        };

        let mut groups: HashMap<String, f64> = HashMap::new();
        for row in &rows {
            let dv = row.get(&date_col).and_then(|v| v.as_str()).unwrap_or("");
            let key = match get_period_key(dv, &period) {
                Some(k) => k,
                None => continue,
            };
            if let Some(v) = parse_row_value(row, &metric_col) {
                *groups.entry(key).or_default() += v;
            }
        }

        let mut sorted: Vec<(String, f64)> = groups.into_iter().collect();
        sorted.sort_by(|a, b| a.0.cmp(&b.0));
        if sorted.len() < 2 {
            return ApiResult::failure("数据点不足");
        }

        let labels: Vec<String> = sorted.iter().map(|(k, _)| k.clone()).collect();
        let values: Vec<f64> = sorted.iter().map(|(_, v)| *v).collect();
        let n = values.len();

        // MA3
        let ma: Vec<Option<f64>> = (0..n)
            .map(|i| {
                if i < 2 {
                    None
                } else {
                    Some((values[i - 2] + values[i - 1] + values[i]) / 3.0)
                }
            })
            .collect();

        // MoM
        let mom: Vec<Option<f64>> = (0..n)
            .map(|i| {
                if i == 0 || values[i - 1] == 0.0 {
                    None
                } else {
                    Some((values[i] - values[i - 1]) / values[i - 1].abs() * 100.0)
                }
            })
            .collect();

        // YoY (monthly only)
        let yoy: Vec<Option<f64>> = if period == "month" && n > 12 {
            (0..n)
                .map(|i| {
                    if i < 12 || values[i - 12] == 0.0 {
                        None
                    } else {
                        Some((values[i] - values[i - 12]) / values[i - 12].abs() * 100.0)
                    }
                })
                .collect()
        } else {
            vec![None; n]
        };

        // Linear regression
        let mean_x = (n - 1) as f64 / 2.0;
        let mean_y = values.iter().sum::<f64>() / n as f64;
        let mut num = 0.0;
        let mut den = 0.0;
        for i in 0..n {
            let dx = i as f64 - mean_x;
            num += dx * (values[i] - mean_y);
            den += dx * dx;
        }
        let slope = if den != 0.0 { num / den } else { 0.0 };
        let intercept = mean_y - slope * mean_x;
        let trend: Vec<f64> = (0..n).map(|i| intercept + slope * i as f64).collect();

        let forecast_labels = next_period_keys(&labels[n - 1], &period, 3);
        let forecast_values: Vec<f64> =
            (0..3).map(|i| intercept + slope * (n + i) as f64).collect();

        ApiResult::success(TimeseriesResult {
            labels,
            values,
            ma,
            mom,
            yoy,
            trend,
            forecast_labels,
            forecast_values,
        })
    })
    .await
    .unwrap_or_else(|e| ApiResult::failure(format!("spawn_blocking error: {e}")))
}

#[tauri::command]
pub async fn compute_deciles(rows_json: String, metric_col: String) -> ApiResult<DecileResult> {
    tokio::task::spawn_blocking(move || {
        let rows: Vec<HashMap<String, serde_json::Value>> = match serde_json::from_str(&rows_json) {
            Ok(r) => r,
            Err(e) => return ApiResult::failure(format!("JSON解析失败: {e}")),
        };

        let mut nums: Vec<f64> = rows
            .iter()
            .filter_map(|r| parse_row_value(r, &metric_col))
            .filter(|n| !n.is_nan())
            .collect();

        if nums.len() < 10 {
            return ApiResult::failure("数据点不足（至少需要10个）");
        }

        nums.sort_by(|a, b| a.partial_cmp(b).unwrap());
        let bucket_size = nums.len() / 10;

        let mut labels = Vec::new();
        let mut counts = Vec::new();
        let mut sums = Vec::new();
        let mut avgs = Vec::new();
        let mut ranges = Vec::new();

        for i in 0..10 {
            let start = i * bucket_size;
            let end = if i == 9 {
                nums.len()
            } else {
                (i + 1) * bucket_size
            };
            let bucket = &nums[start..end];
            let sum: f64 = bucket.iter().sum();

            labels.push(format!("D{:02}", i + 1));
            counts.push(bucket.len());
            sums.push(sum);
            avgs.push(if bucket.len() > 0 {
                sum / bucket.len() as f64
            } else {
                0.0
            });
            ranges.push(format!(
                "{}–{}",
                fmt_compact(bucket[0]),
                fmt_compact(bucket[bucket.len() - 1])
            ));
        }

        ApiResult::success(DecileResult {
            labels,
            counts,
            sums,
            avgs,
            ranges,
        })
    })
    .await
    .unwrap_or_else(|e| ApiResult::failure(format!("spawn_blocking error: {e}")))
}

#[tauri::command]
pub async fn compute_clusters(
    rows_json: String,
    metric_cols: Vec<String>,
    k: usize,
    x_col: Option<String>,
    y_col: Option<String>,
) -> ApiResult<ClusterResult> {
    tokio::task::spawn_blocking(move || {
        let rows: Vec<HashMap<String, serde_json::Value>> = match serde_json::from_str(&rows_json) {
            Ok(r) => r,
            Err(e) => return ApiResult::failure(format!("JSON解析失败: {e}")),
        };

        let col_x = x_col.unwrap_or_else(|| metric_cols.get(0).cloned().unwrap_or_default());
        let col_y = y_col.unwrap_or_else(|| {
            metric_cols
                .get(1)
                .or(metric_cols.get(0))
                .cloned()
                .unwrap_or_default()
        });
        if col_x.is_empty() || col_y.is_empty() {
            return ApiResult::failure("缺少指标列");
        }

        let k = k.max(1).min(10);
        let mut points: Vec<(f64, f64, String)> = Vec::new();
        for row in &rows {
            if let (Some(x), Some(y)) = (parse_row_value(row, &col_x), parse_row_value(row, &col_y))
            {
                let label = row
                    .values()
                    .next()
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string();
                points.push((x, y, label));
            }
        }
        if points.len() < k {
            return ApiResult::failure("数据点不足");
        }

        // Init centroids
        let step = points.len() / k;
        let mut centroids: Vec<(f64, f64)> = (0..k)
            .map(|i| {
                let idx = (i * step).min(points.len() - 1);
                (points[idx].0, points[idx].1)
            })
            .collect();

        // Iterate
        let mut assignments = vec![0usize; points.len()];
        for _ in 0..20 {
            let new_assign: Vec<usize> = points
                .iter()
                .map(|(px, py, _)| {
                    let mut min_dist = f64::MAX;
                    let mut min_c = 0;
                    for (c, (cx, cy)) in centroids.iter().enumerate() {
                        let d = (px - cx).powi(2) + (py - cy).powi(2);
                        if d < min_dist {
                            min_dist = d;
                            min_c = c;
                        }
                    }
                    min_c
                })
                .collect();

            if new_assign == assignments {
                break;
            }
            assignments = new_assign;

            let mut sums: Vec<(f64, f64, usize)> = vec![(0.0, 0.0, 0); k];
            for (i, (px, py, _)) in points.iter().enumerate() {
                sums[assignments[i]].0 += px;
                sums[assignments[i]].1 += py;
                sums[assignments[i]].2 += 1;
            }
            centroids = sums
                .iter()
                .map(|(sx, sy, c)| {
                    if *c > 0 {
                        (sx / *c as f64, sy / *c as f64)
                    } else {
                        (0.0, 0.0)
                    }
                })
                .collect();
        }

        let cluster_points: Vec<ClusterPoint> = points
            .iter()
            .enumerate()
            .map(|(i, (x, y, label))| ClusterPoint {
                x: *x,
                y: *y,
                cluster: assignments[i],
                label: label.clone(),
            })
            .collect();

        ApiResult::success(ClusterResult {
            points: cluster_points,
            centroids,
            col_x,
            col_y,
        })
    })
    .await
    .unwrap_or_else(|e| ApiResult::failure(format!("spawn_blocking error: {e}")))
}
